[Misc] Add OpenTelemetry support

ronensc opened the pull request:
This PR adds basic support for OpenTelemetry distributed tracing.
It includes changes to enable tracing functionality and improve monitoring capabilities.

I've also added a markdown with print-screens to guide users how to use this feature. You can find it [here](https://github.com/ronensc/vllm/blob/otel/examples/production_monitoring/Otel.md)

FIX #3789

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

ronensc committed:
Add OpenTelemetry support

PRATIBHA-Moogi commented:
@rkooo567 can you please review this PR raised by Ronen. Thanks.

rkooo567 commented:
yep will do tmrw!

ronensc commented:
Rebased to fix merge conflicts.

gyliu513 commented on vllm/engine/llm_engine.py:
```diff
@@ -885,3 +904,43 @@ def list_loras(self) -> List[int]:
+            ttft = metrics.first_token_time - metrics.arrival_time
+            e2e_time = now - seq_group.metrics.arrival_time
+            # attribute names are based on
+            # https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/llm-spans.md
```
Thanks for the adoption of otel llm semantic convetion, this is great!

saurabhjha1 replied:
Great. also do we have a holistic view of different observability metrics in vLLM?

I see another effort here: https://github.com/vllm-project/vllm/issues/5041

ronensc replied:
We have this issue: https://github.com/vllm-project/vllm/issues/3616#issuecomment-2030858781

ronensc committed:
Create a trace only when trace_context is not None

nirga commented on vllm/engine/llm_engine.py:
```diff
@@ -974,3 +990,45 @@ def list_loras(self) -> List[int]:
+            e2e_time = now - seq_group.metrics.arrival_time
+            # attribute names are based on
+            # https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/llm-spans.md
+            seq_span.set_attribute("gen_ai.response.model",
```
Can I suggest using the constants from https://github.com/traceloop/openllmetry/blob/main/packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv/ai/__init__.py
?
We keep them up to date with the decisions we're making in the otel semantic conventions working group

ronensc replied:
Thanks for your suggestion! I've added a dependency on `opentelemetry-semantic-conventions-ai` and extended `SpanAttributes` with the missing constants.

ronensc committed:
Use constants for span attribute names

rkooo567 commented on vllm/engine/llm_engine.py:
```diff
@@ -974,3 +990,48 @@ def list_loras(self) -> List[int]:
+            return
+        arrival_time_nano_seconds = int(seq_group.metrics.arrival_time * 1e9)
+
+        with tracer.start_as_current_span(
```
QQ: what's the e2e overhead of this API? Can you measure and lmk for prompt length 512 output length 128?

rkooo567 reviewed (commented):
Code itself looks good to me. My main concern is that we should probably not include this by default? This requires to setup Jaeger (or other tracing tools), and this seems pretty heavy requirements for every user. 

That said, it'd require 3 changes. 
- remove the requirements from requirements-common.txt 
- init_trace should try importing it and becomes no-op when packages are not downloaded.
- Fail the request if tracing is not enabled.

cc @simon-mo for thoughts!

simon-mo commented:
It needs to be opt-in (i think it is already?), similar to TGI:
> text-generation-inference is instrumented with distributed tracing using OpenTelemetry. You can use this feature by setting the address to an OTLP collector with the --otlp-endpoint argument.

ronensc committed:
Remove required dependency on opentelemetry

SemMulder commented on vllm/tracing.py:
```diff
@@ -0,0 +1,31 @@
+from opentelemetry.trace import set_tracer_provider
+
+
+def init_tracer(instrumenting_module_name):
+    trace_provider = TracerProvider()
+
+    # The endpoint of OTLPSpanExporter is set from envvars:
+    #  OTEL_EXPORTER_OTLP_ENDPOINT
+    #  OTEL_EXPORTER_OTLP_TRACES_ENDPOINT
+    trace_provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter()))
+    set_tracer_provider(trace_provider)
+
+    tracer = trace_provider.get_tracer(instrumenting_module_name)
+    return tracer
```
Instead the SpanProcesser should be setup as part of the initialization process of the OpenAI-compatible server.

ronensc replied:
Thanks. Good catch. Has introducing the `--otlp-endpoint` flag solved this issue?

SemMulder replied:
> Has introducing the --otlp-endpoint flag solved this issue?

Not entirely :). The SpanProcesser is still set up in `init_tracer` via `LLMEngine.__init__`, which still results in a SpanProcessor being added when using VLLM as a library.

The following chunk:

```python
    trace_provider = TracerProvider()

    # The endpoint of OTLPSpanExporter is set from envvars:
    #  OTEL_EXPORTER_OTLP_ENDPOINT
    #  OTEL_EXPORTER_OTLP_TRACES_ENDPOINT
    trace_provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter()))
    set_tracer_provider(trace_provider)
```
should be moved to the initialization specific to the OpenAI-compatible server (e.g. [here](https://github.com/ronensc/vllm/blob/7f5cd5a7ac4f760bfc87f56f01153d691edb7496/vllm/entrypoints/api_server.py#L109)).

SemMulder replied:
`init_tracer` is only called when `--otlp-endpoint` is set, of course. But we are using VLLM as a library and would also like to use the OpenTelemetry instrumentation introduced here :).

ronensc replied:
To configure the `otlp_endpoint` flag for offline inferencing, use the following syntax:
```python
from vllm import LLM, SamplingParams

llm = LLM(
	model="facebook/opt-125m", 
	disable_log_stats=False, 
	otlp_endpoint="http://172.17.0.2:4318/v1/traces",
)
```
However, it's important to note that currently, no traces will be reported because `trace_context` is set to None:
https://github.com/vllm-project/vllm/pull/4687/files#diff-c89ac25bd066e936e80260d21be63c7d2379cfedc371a9ff288fb5ba02ae1350R1003-R1004

I'll work on implementing a fix for it.

ronensc committed:
Add --otlp_endpoint flag

rkooo567 commented on examples/production_monitoring/Otel.md:
should we update the flag here?

rkooo567 commented on examples/production_monitoring/Otel.md:
can you update this requires dependencies (and add pip install example?)

rkooo567 reviewed (commented):
Looks pretty good to me. Is it possible to also add a simple unit test on top of the final todo? I will approve it after that!

ronensc committed:
Write a one-time warning log for requests with a tracing header when tracing is disabled

ronensc committed:
Relax condition

This should allow tracing in offline inference mode.

ronensc committed:
Decouple tracing for log stats

ronensc committed:
Rename --otlp-endpoint -> --otlp-traces-endpoint

ronensc committed:
Update Otel.md

ronensc committed:
Support grpc exporter and make it the default

ronensc committed:
Update Otel.md

ronensc committed:
Rename OTEL_EXPORTER_OTLP_ENDPOINT -> OTEL_EXPORTER_OTLP_TRACES_ENDPOINT

ronensc committed:
Change exporter in dummy_client to grpc

ronensc committed:
Add --otlp-traces-endpoint flag to benchmark_latency.py

ronensc committed:
Add e2e test for tracing

ronensc committed:
Add tracing to CI

ronensc commented:
Since previous status:
1. Decoupled tracing from `log_stats`.
1. Added support for grpc exporter.
1. Added a one-time warning log for requests with a tracing header when tracing is disabled.
1. Updated the markdown guide.
1. Added a simple e2e unit test.
1. Measured the e2e overhead of sending traces:

Baseline: Without `--otlp-traces-endpoint`
```
$ OTEL_EXPORTER_OTLP_TRACES_INSECURE=true python3 benchmarks/benchmark_latency.py \
	--input-len 512 \
	--output-len 128 2>&1

...
Avg latency: 0.5273534125333412 seconds
10% percentile latency: 0.5149867414999904 seconds
25% percentile latency: 0.5197465127499754 seconds
50% percentile latency: 0.5262749200000485 seconds
75% percentile latency: 0.5300818937500082 seconds
90% percentile latency: 0.5346117969000375 seconds
```

With `--otlp-traces-endpoint`
```
$ OTEL_EXPORTER_OTLP_TRACES_INSECURE=true python3 benchmarks/benchmark_latency.py \
	--input-len 512 \
	--output-len 128 \
	--otlp-traces-endpoint grpc://$JAEGER_IP:4317 2>&1

...
Avg latency: 0.5384878202333273 seconds
10% percentile latency: 0.5267181372999744 seconds
25% percentile latency: 0.5291085377500337 seconds
50% percentile latency: 0.5376283054999362 seconds
75% percentile latency: 0.5437845339999967 seconds
90% percentile latency: 0.5500757966999685 seconds
```


I ran these experiments on my laptop using WSL2, with Jaeger running in Docker.
GPU: NVIDIA RTX A4000

ronensc commented:
> Lmk when it is ready!

ready :)

ronensc commented:
It appears that the CI is failing due to an issue with tracing. I'll investigate and add a fix.

ronensc committed:
Pass trace headers instead of trace context

When --engine-use-ray is set, the server cannot pass otel Context as its immutable.

ronensc committed:
Add is_tracing_enabled()

When --engine-use-ray is set, the server cannot access engine.tracer.
Instead, we expose whether tracing is enabled through is_tracing_enabled().

ronensc committed:
Handle passing None to extract_trace_context

rkooo567 commented on vllm/tracing.py:
```diff
@@ -0,0 +1,102 @@
+
+def init_tracer(instrumenting_module_name: str,
+                otlp_traces_endpoint: str) -> Optional[Tracer]:
+    trace_provider = TracerProvider()
```
Can you assert open telemetry is imported here and raise an exception to pip install deps?

ronensc replied:
Done. Additionally, please note that there is a similar check in the configuration for validating CLI arguments: 
https://github.com/vllm-project/vllm/blob/9fa949e268805dbb411992b79f25c9a6b7fdcb59/vllm/config.py#L1371-L1374

rkooo567 reviewed (approved):
Looks pretty good. there are merge conflicts

ronensc committed:
Assert otel is installed in init_tracer()

ronensc commented on vllm/config.py:
```diff
@@ -1362,6 +1363,17 @@ def __post_init__(self):
+    """Configuration for observability."""
+    otlp_traces_endpoint: Optional[str] = None
+
+    def __post_init__(self):
+        if not is_otel_installed() and self.otlp_traces_endpoint is not None:
+            raise ValueError("OpenTelemetry packages must be installed before "
+                             "configuring 'otlp_traces_endpoint'")
```
Done. Additionally, please note that there is a similar check in the configuration for validating CLI arguments: 
https://github.com/vllm-project/vllm/blob/9fa949e268805dbb411992b79f25c9a6b7fdcb59/vllm/config.py#L1371-L1374

ronensc reviewed (commented):
@rkooo567 thanks for your review!
I have rebased once again to resolve the merge conflicts. If the CI passes, I believe the PR can be merged.

ronensc committed:
Fix bug itroduced by git rebase

ronensc committed:
Fix lint error

AndSonder commented:
@rkooo567  
Hello! I hope this message finds you well. I have a question about adding additional metrics. I came across the documentation here: https://docs.vllm.ai/en/latest/serving/metrics.html, which introduces various metrics. I believe incorporating these could greatly help in diagnosing and addressing performance bottlenecks in the model. Would it be possible for you to provide some guidance or suggestions on how to add these metrics? I would sincerely appreciate your help. Thank you so much!
